juniper/src/parser/utils.rs

```diff
@@ -8,99 +8,129 @@ pub struct SourcePosition {
     col: usize,
 }
 
-/// Data structure used to wrap items with start and end markers in the input source
-///
-/// A "span" is a range of characters in the input source, starting at the
-/// character pointed by the `start` field and ending just before the `end`
-/// marker.
-#[derive(Debug, Clone, PartialEq, Eq, Hash, Copy)]
-pub struct Spanning<T> {
-    /// The wrapped item
-    pub item: T,
-
-    /// Start position of the item
+/// Range of characters in the input source, starting at the character pointed by the `start` field
+/// and ending just before the `end` marker.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct Span {
+    /// Start position of this [`Span`].
     pub start: SourcePosition,
 
-    /// End position of the item
+    /// End position of this [`Span`].
     ///
-    /// This points to the first source position _after_ the wrapped item.
+    /// > __NOTE__: This points to the first source position __after__ this [`Span`].
     pub end: SourcePosition,
 }
 
-impl<T> Spanning<T> {
+impl Span {
     #[doc(hidden)]
-    pub fn zero_width(pos: &SourcePosition, item: T) -> Spanning<T> {
-        Spanning {
-            item,
-            start: *pos,
-            end: *pos,
+    #[inline]
+    pub fn zero_width(pos: SourcePosition) -> Self {
+        Self {
+            start: pos,
+            end: pos,
         }
     }
 
     #[doc(hidden)]
-    pub fn single_width(pos: &SourcePosition, item: T) -> Spanning<T> {
-        let mut end = *pos;
+    #[inline]
+    pub fn single_width(pos: SourcePosition) -> Self {
+        let mut end = pos;
         end.advance_col();
 
-        Spanning {
-            item,
-            start: *pos,
-            end,
-        }
+        Self { start: pos, end }
     }
 
     #[doc(hidden)]
-    pub fn start_end(start: &SourcePosition, end: &SourcePosition, item: T) -> Spanning<T> {
-        Spanning {
-            item,
-            start: *start,
-            end: *end,
+    #[inline]
+    pub fn unlocated() -> Self {
+        Self {
+            start: SourcePosition::new_origin(),
+            end: SourcePosition::new_origin(),
         }
     }
+}
+
+/// Data structure used to wrap items into a [`Span`].
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct Spanning<T> {
```

Comment:
In the planned follow-up PR/rebase, which adds spanning to arguments and LookAheads, this struct gets a second generic argument `<Sp = Span>`, so that the span can be a reference in some contexts (one-word copy instead of six words).

I was a bit unsure whether to include that change in this PR instead, because I got the idea after this refactor was done.

Reply:
@audunhalland it would be OK to have in the follow-up PR.